pkg/apis/serving/v1alpha1/kfservice_framework.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2019 kubeflow.org.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+	"log"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+type FrameworkHandler interface {
+	CreateModelServingContainer(modelName string) *v1.Container
+	Validate() error
+}
+
+const (
+	// ExactlyOneModelSpecViolatedError is a known error message
+	ExactlyOneModelSpecViolatedError = "Exactly one of [Custom, Tensorflow, ScikitLearn, XGBoost] must be specified in ModelSpec"
+	// AtLeastOneModelSpecViolatedError is a known error message
+	AtLeastOneModelSpecViolatedError = "At least one of [Custom, Tensorflow, ScikitLearn, XGBoost] must be specified in ModelSpec"
+)
+
+func (m *ModelSpec) CreateModelServingContainer(modelName string) *v1.Container {
+	handler, err := makeHandler(m)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	return handler.CreateModelServingContainer(modelName)
+}
+
+func (m *ModelSpec) Validate() error {
```

Comment:
It may not be necessary right now, but we should consider that ValidateCreate and ValidateUpdate are the two primitives that need to be supported for webhooks. I don't think we have any restrictions on update (and I really want to keep it that way), but it's something to keep in mind.

Reply:
Currently ValidateUpdate ignores the old object and just calls Validate - which equates to: "you validate yourself"... If we change that (and I agree will be nice not to have to), we will have to add a a method on the interface that checks if the fields that changed are mutable by passing in old and new to validate... so in that sense an extra fn.. so going to keep the method as it is for now.

Reply:
SGTM